numba/np/linalg.py

```diff
@@ -521,39 +520,52 @@ def check_args(a, b):
     return builder.load(out)
 
 
-@glue_lowering(np.dot, types.Array, types.Array)
-def dot_2(context, builder, sig, args):
+@overload(np.dot)
+@overload(operator.matmul)
+def dot_2(left, right):
     """
     np.dot(a, b)
     a @ b
     """
-    ensure_blas()
-
-    with make_contiguous(context, builder, sig, args) as (sig, args):
-        ndims = [x.ndim for x in sig.args[:2]]
-        if ndims == [2, 2]:
-            return dot_2_mm(context, builder, sig, args)
-        elif ndims == [2, 1]:
-            return dot_2_mv(context, builder, sig, args)
-        elif ndims == [1, 2]:
-            return dot_2_vm(context, builder, sig, args)
-        elif ndims == [1, 1]:
-            return dot_2_vv(context, builder, sig, args)
-        else:
-            assert 0
+    if isinstance(left, types.Array) and isinstance(right, types.Array):
+        @intrinsic
+        def _impl(typingcontext, left, right):
+            def _dot2_codegen(context, builder, sig, args):
+                ensure_blas()
+
+                with make_contiguous(context, builder, sig, args) as (sig, args):
+                    ndims = [x.ndim for x in sig.args[:2]]
+                    if ndims == [2, 2]:
+                        return dot_2_mm(context, builder, sig, args)
+                    elif ndims == [2, 1]:
+                        return dot_2_mv(context, builder, sig, args)
+                    elif ndims == [1, 2]:
+                        return dot_2_vm(context, builder, sig, args)
+                    elif ndims == [1, 1]:
+                        return dot_2_vv(context, builder, sig, args)
+                    else:
+                        assert 0
+
+            return signature(types.Array, left, right), _dot2_codegen
```

Comment:
```
            return signature(types.Array, left, right), _dot2_codegen
                             ^^^^^^^^^^^
```
needs to be a instance of `types.Array`. you can get it from the logic in npydecl.py